File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    
    const visited = Array.from({length: n}, () => Array(m).fill(false));
    visited[0][0] = true;
    
    const queue = [[0,0,1]];    
    let head = 0;
    
    const dx = [1, -1, 0, 0];
    const dy = [0, 0, 1, -1];

    while (head < queue.length) {

        const [cx, cy, step] = queue[head++];
       
        for (let i = 0 ; i < 4 ; i++) {
            const [nx, ny] = [cx+dx[i], cy+dy[i]];
             if (nx === n - 1 && ny === m - 1) {
                return step + 1;
            }

            const isInside = 0 <= nx && nx < n && 0 <= ny && ny < m;

            if (isInside && maps[nx][ny] === 1 && !visited[nx][ny]) {
                visited[nx][ny] = true;
                queue.push([nx, ny, step + 1]);
            }
        }
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    
    answer = 0

    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]

    n = len(maps)
    m = len(maps[0])
    

    queue = deque([(0, 0, 1)])
    maps[0][0] = 0
    
    while queue:
        x, y, c = queue.popleft()
        
        if x == n - 1 and y == m - 1:
            return c
        
        for i in range(4):        
            nx, ny = x + dx[i], y + dy[i]
            
            if nx < 0 or nx >= n or ny < 0 or ny >= m or maps[nx][ny] == 0:
                continue
            
            maps[nx][ny] = 0
            queue.append((nx, ny, c + 1))
    
    return -1